source/vsignal/sink.d

```diff
@@ -0,0 +1,128 @@
+module vsignal.sink;
+
+import vsignal.signal;
+import vsignal.slot : Slot, slot_connect = connect;
+
+Connection connect(alias pred, F)(auto ref Sink!F sink)
+{
+	sink.disconnect!pred;
+
+	import core.lifetime : move;
+
+	with (sink)
+	{
+		Slot!F call;
+		call.slot_connect!pred;
+		signal.calls ~= call.move();
+
+		Slot!(void delegate(void* signal) @safe pure nothrow) conn;
+		conn.slot_connect!(release!(pred, F))();
+
+		return Connection(conn.move(), signal);
+	}
+}
+
+Connection connect(alias pred, T, F)(auto ref Sink!F sink, ref T instance)
+{
+	sink.disconnect!pred(instance);
+
+	import core.lifetime : move;
+
+	with (sink)
+	{
+		Slot!F call;
+		call.slot_connect!pred(instance);
+		signal.calls ~= call.move();
+
+		Slot!(void delegate(void* signal) @safe pure nothrow) conn;
+		conn.slot_connect!(release!(pred, T, F))(instance);
+
+		return Connection(conn.move(), signal);
+	}
+}
+
+void disconnect(alias pred, F)(auto ref Sink!F sink)
+{
+	import std.algorithm : remove, SwapStrategy;
+	import std.algorithm : countUntil;
+
+	Slot!F call;
+	call.slot_connect!pred;
+
+	with (sink)
+	{
+		const index = signal.calls.countUntil(call);
+
+		if (index > -1)
+			signal.calls = signal.calls.remove!(SwapStrategy.unstable)(index);
+	}
+}
+
+void disconnect(alias pred, T, F)(auto ref Sink!F sink, ref T instance)
+{
+	import std.algorithm : remove, SwapStrategy;
+	import std.algorithm : countUntil;
+
+	Slot!F call;
+	call.slot_connect!pred(instance);
+
+	with (sink)
+	{
+		const index = signal.calls.countUntil(call);
+
+		if (index > -1)
+			signal.calls = signal.calls.remove!(SwapStrategy.unstable)(index);
+	}
+}
+
+void disconnect(T, F)(auto ref Sink!F sink, ref T instance)
+{
+	import std.algorithm : remove, SwapStrategy;
+
+	with (sink)
+	{
+		signal.calls = signal.calls.remove!(call => call.payload is &instance , SwapStrategy.unstable);
+	}
+}
+
+void disconnect(F)(auto ref Sink!F sink)
+{
+	import std.algorithm : remove, SwapStrategy;
+
+	with (sink)
+	{
+		signal.calls = [];
+	}
+}
```

Comment:
True, those were implemented lastly and I completely forgot about it. Just went with the rest of the design.

Reply:
Well never mind, this causes issues with the UFCS. If the type has a function declared in its scope the compiler will never fall back to a global alternative. This means that if **Sink** has those `disconnect` functions declared, the user will never be able to do `sink.disconnect!pred` since there is no overload declared in **Sink** that matches that signature. The only way is to call the function which I absolutely do not want. I will keep this as is.